utils/src/ipc.rs

```diff
@@ -360,3 +374,40 @@ pub fn get_cache_path() -> Result<PathBuf, String> {
 
     Ok(cache_path)
 }
+
+/// We make sure the Stream is always set to blocking mode
+///
+/// * `tries` -  how many times to attempt the connection
+/// * `interval` - how long to wait between attempts, in milliseconds
+pub fn connect_to_socket(addr: &PathBuf, tries: u8, interval: u64) -> Result<UnixStream, String> {
+    //Make sure we try at least once
+    let tries = if tries == 0 { 1 } else { tries };
+    let mut error = None;
+    for _ in 0..tries {
+        match UnixStream::connect(addr) {
+            Ok(socket) => {
+                if let Err(e) = socket.set_nonblocking(false) {
+                    return Err(format!("Failed to set blocking connection: {e}"));
+                }
+                #[cfg(debug_assertions)]
+                let timeout = Duration::from_secs(30); //Some operations take a while to respond in debug mode
+                #[cfg(not(debug_assertions))]
+                let timeout = Duration::from_secs(5);
+
+                if let Err(e) = socket.set_read_timeout(Some(timeout)) {
+                    return Err(format!("failed to set read timeout for socket: {e}"));
+                }
+
+                return Ok(socket);
+            }
+            Err(e) => error = Some(e),
+        }
+        std::thread::sleep(Duration::from_millis(interval));
+    }
+    let error = error.unwrap();
+    if error.kind() == std::io::ErrorKind::NotFound {
+        return Err("Socket file not found. Are you sure swww-daemon is running?".to_string());
```

Comment:
Alright, now we have a new problem. This message will be printed every time we try to initialize the daemon when there wasn't a previous socket file there. We do not want that.

So, I'd recommend changing this function to return a `std::io::Result<UnixStream>`. Then, all the errors we are currently using `format!` on can be simplified. For example:

```rust
if let Err(e) = socket.set_nonblocking(false) {
    return Err(format!("Failed to set blocking connection: {e}"));
}
```
Can become just
```rust
socket.set_nonblocking(false)?;
```
And so on.

After that, you will need to fix the error handling in every place (4) this function is called. For most of them, I think you can just do `.map(|e| format!("failed to connect to socket: {e}")?;` on the function call.

There's one specific case, in the client's main function, where you will have to check if the error is `std::io::ErrorKind::NotFound`, and return a special message if it is.

Reply:
> This message will be printed every time we try to initialize the daemon when there wasn't a previous socket file there.

The ``is_daemon_running`` function is only called if the socket file exists. If it doesn't, that entire block gets skipped over. The error is also suppressed if the file exists but there's no daemon listening.

I can still refacttor the error handling if you wish, but that feels like it would be out of scope for this PR...?

Reply:
Ah, I see. Indeed, you are right. I was seeing it print an error because I was testing it the wrong way.